tests/compile-fail/permissive_provenance_null.rs

```diff
@@ -0,0 +1,14 @@
+// compile-flags: -Zmiri-permissive-provenance
+#![feature(strict_provenance)]
+
+use std::ptr;
+
+fn main() {
+    let x = 5;
+    let x_ptr = &x as *const i32;
+    let x_usize = x_ptr.expose_addr();
+
+    unsafe {
+        let _a = *ptr::from_exposed_addr::<i32>(0).with_addr(x_usize); //~ ERROR attempting a read access using
```

Comment:
This test is suspicious. Why should we reject this?
I think the `from_exposed_addr` spec allows this.

Reply:
I was following the LangRed when doing this https://llvm.org/docs/GetElementPtr.html#id19.

I could make this code work, but I need to fix the ptr_get_alloc logic to only do the find best allocation logic for `Tag::Wildcard` (and not also `Tag::None`), and that breaks the null case.

Reply:
Sorry, was a bit confused when I mention `Tag::None`. Something is letting me allocation hop even with a tag of `None`, and I'm not sure where that is.

We can either reject this according to the LLVM LangRef or accept it alongside the other inconsistencies, but I need to look into this part of the code a bit more.

Reply:
I am not sure LLVM LangRef says to reject this.
After all, this *could* have picked up the right provenance, the one of `x_ptr`.

> Something is letting me allocation hop even with a tag of None, and I'm not sure where that is.

I don't understand what you mean by this.

Reply:
Ah I saw an email notification for your other comment, but cannot find it on GH.

> I was following the LangRed when doing this https://llvm.org/docs/GetElementPtr.html#id19.

LangRef has no clue how to specify ptr-int roundtrips, so I think we should ignore it. Rather, the relevant documentation is `from_exposed_addr` in Rust.

Reply:
> I could make this code work, but I need to fix the ptr_get_alloc logic to only do the find best allocation logic for Tag::Wildcard (and not also Tag::None), and that breaks the null case.

Indeed `None` should never do "find best allocation" logic, that is the entire point of distinguishing `Wildcard` and `None`.

Reply:
Alright, I've fixed the fact that cast and transmute weren't treated differently. The code here still fails the null test-case I've added though, as the assert in rustc requires that pointers with address 0 have no provenance, and offsetting a pointer with no provenance produces another pointer with no provenance.

Can I remove this assert in rustc / is that sound given what the reset of const-eval/the compiler assumes? Or should I hardcode something in the implementation of `wrapping_offset` to make this work?

Reply:
> Can I remove this assert in rustc / is that sound given what the reset of const-eval/the compiler assumes?

Yes please remove that assert. I did not quite realize the consequences it has.
We can already produce pointers with NULL address and `AllocId` provenance via `ptr.with_addr(0)`; I don't see why creating them also via `from_exposed_addr(0)` would cause any new problems.